### PR TITLE
feat: Add `Mutex`

### PR DESCRIPTION
**Pull request by scottbot95**
Implement async mutex

### CODE REVIEW

File: screeps-async/src/sync/mutex.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            data: UnsafeCell::new(val),
+            wakers: UnsafeCell::new(VecDeque::new()),
+        }
+    }
+
+    /// Acquire the mutex.
+    ///
+    /// Returns a guard that release the mutex when dropped
+    pub async fn lock(&self) -> MutexGuard<'_, T> {
+        MutexLockFuture::new(self).await
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs]: This is an anti-pattern; if you have a custom future type, make this a sync function and return that custom type.

File: screeps-async/src/sync/mutex.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+///     *val = 1;
+/// }).detach();
+/// ```
+pub struct Mutex<T> {
+    /// Whether the mutex is currently locked.
+    state: AtomicBool,
+    /// Wrapped value
+    data: UnsafeCell<T>,
+    /// Queue of futures to wake when a lock is released
+    wakers: UnsafeCell<VecDeque<Rc<UnsafeCell<Waker>>>>,
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs]: Why not just `UnsafeCell<Exclusive<VecDequeue<Waker>>>`? Now the Mutex is send + sync and you avoided the Rc/RefCell overhead.

File: screeps-async/src/sync/mutex.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+pub struct MutexGuard<'a, T> {
+    lock: &'a Mutex<T>,
+}
+
+impl<'a, T> MutexGuard<'a, T> {
+    fn new(lock: &'a Mutex<T>) -> Self {
+        Self { lock }
+    }
+}
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs:95]: Have you considered adding an `async fn unlock_fair(self)` that forces a yield back to the executor after unlocking?

File: screeps-async/src/sync/mutex.rs (hunk 2 above)
Comment (khoover) [reply screeps-async/src/sync/mutex.rs]: Disregard  forgot the blanket !Sync for UnsafeCell. You'd need to manually add Send + Sync impls for the Mutex.

File: screeps-async/src/sync/mutex.rs (hunk 1 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs]: Ah interesting. I'd seen some example doing basically this which makes some sense to me as the return type is more obvious in this case. However I can see how the extra layer of wrapping would be considered an anti-pattern.

File: screeps-async/src/sync/mutex.rs (hunk 2 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs]: I was not aware of `Exclusive`. Looks like that's a nightly only feature though and I don't want to require that of users.

Also the inner Rc<UnsafeCell<Waker>> is used so I can share a mutable reference to the waker between here (where it will be called from) and the Future implementation (so that it can be updated via `Waker::clone_from`).

Good catch on the Send + Sync, I will add those.

File: screeps-async/src/sync/mutex.rs (hunk 1 above)
Comment (khoover) [reply screeps-async/src/sync/mutex.rs]: It's the same as Iterator-implementing types, if you can return the exact struct over an `impl Iterator` it probably makes more sense to return the exact struct.

File: screeps-async/src/sync/mutex.rs (hunk 3 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs:95]: Hmmm I had not thought about that. What that be a function on the guard then? Or perhaps an associated function on Mutex that consumes the guard? I'm envisioning something like this (probably would make regular `unlock` public too and rework it to also consume the guard)

```rs
impl<T> Mutex<T> {
    pub async fn unlock_fair(guard: MutexGuard<'_, T>) {
        drop(guard);
        crate::yield_now().await;
    }
}


```

File: screeps-async/src/sync/mutex.rs (hunk 2 above)
Comment (khoover) [reply screeps-async/src/sync/mutex.rs]: Is it still nightly-only? Bother, you'd think such a simple thing would have stabilized already given the benefits.

Also, I just remembered the other problem with only popping a single `Waker`; what if the Future that added the Waker has since been dropped/forgotten, so now it never takes the lock or wakes any of the other pending wakers? If you drain the queue each time, then I think the updating point becomes moot.

File: screeps-async/src/sync/mutex.rs (hunk 3 above)
Comment (khoover) [reply screeps-async/src/sync/mutex.rs:95]: I'd say make it a method on the guard, making it an associated function of the Mutex feels odd to me.

File: screeps-async/src/sync/mutex.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    /// Try to acquire the mutex.
+    ///
+    /// If the mutex could not be acquired at this time return [`None`], otherwise
+    /// returns a guard that will release the mutex when dropped.
+    pub fn try_lock(&self) -> Option<MutexGuard<'_, T>> {
+        self.state
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Acquire)
+            .ok()?;
+        Some(MutexGuard::new(self))
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs]: Fun fact: the way AtomicBool works, you can rewrite this as
```rust
(!self.state
    .swap(true, Ordering::Acquire))
    .then(|| MutexGuard::new(self))
```

File: screeps-async/src/sync/mutex.rs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    /// Consumes the mutex, returning the underlying data
+    pub fn into_inner(self) -> T {
+        self.data.into_inner()
+    }
+
+    fn unlock(&self) {
+        self.state.swap(false, Ordering::Release);
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs]: Ditto here, no need for `swap`, just `store`.

File: screeps-async/src/sync/mutex.rs (hunk 4 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs]: Oh wow that's neat!

File: screeps-async/src/sync/mutex.rs (hunk 3 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs:95]: Sure that seems reasonable. I just noticed that `std::sync::Mutex` has an associated `unlock` that consumes the guard.

File: screeps-async/src/sync/mutex.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        self.data.into_inner()
+    }
+
+    fn unlock(&self) {
+        self.state.store(false, Ordering::Release);
+        let wakers = unsafe { &mut *self.wakers.get() };
+        for waker in wakers.iter() {
+            waker.wake_by_ref();
+        }
+        *wakers = Vec::new();
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs]: How about replacing these four lines with `wakers.drain(..).for_each(Waker::wake);`?

Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs]: Ah neat. I was unaware of drain. I like that since presumably it will not effect the Vec's capacity so requeueing all the Wakers that are still blocked won't cause churn while growing the new Vec.

File: screeps-async/src/sync/mutex.rs [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+///     *val = 1;
+/// }).detach();
+/// ```
+pub struct Mutex<T> {
+    /// Whether the mutex is currently locked.
+    state: AtomicBool,
+    /// Wrapped value
+    data: UnsafeCell<T>,
+    /// Queue of futures to wake when a lock is released
+    wakers: UnsafeCell<Vec<Waker>>,
```
Comment (khoover) [inline screeps-async/src/sync/mutex.rs:31]: So the way you use `wakers` is actually entirely thread unsafe, e.g. what if one thread is unlocking but another is about to push a waker onto the queue? In light of that, why not skip the `AtomicBool` as well (since this struct is definitely `!Sync`) and instead use a `Cell<bool>`?

**Review by khoover — CHANGES_REQUESTED** [review]

File: screeps-async/src/sync/mutex.rs (hunk 7 above)
Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs:31]: I am aware that it is thread unsafe, but I figured it doesn't matter since there is only one thread in screeps anyway though, right?

As far as `AtomicBool` vs `Cell<bool>`, I suppose I could use a Cell since we never need to read without writing. I don't see any reason to choose one over the other though (presumably the generated bytecode will be almost identical for both)

Comment (khoover) [reply screeps-async/src/sync/mutex.rs:31]: Agreed for a single-threaded context, it being sync or not doesn't matter.

As for AtomicBool vs. Cell<bool>, in the WASM context there's no difference, but there's also no real reason to use atomic values if the mutex won't be shared between threads. Plus it would then be usable in other single-threaded contexts where atomic bools do have an overhead versus plain booleans.

Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs:31]: That's fair. I might change it to Cell then. Although I think while it's generally a good thing to not couple the runtime too tightly to screeps/WASM, ultimately that is the only environment I plan to officially support so I'm not *that* adverse to making decisions off restrictions of that environment.

Also, I plan to implement a `RwLock` at some point as well. Idealistically I'd like them to use very similar implementations and I don't think Cell will work for RwLock since we would need to actually use the previous state to compute the next state. It's not that big a deal if they differ though. It would be nice if I could just use [Cell::update](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.update) but that's also nightly for some reason so I'd like to avoid using it (things seem to stay in nightly forever...). I suppose I could always just `Cell::replace` with some junk value and replace a second time after updating the original which probably isn't really much slower at all compared to however `update` would work.

Comment (khoover) [reply screeps-async/src/sync/mutex.rs:31]: For RwLock, I think you could honestly just use a RefCell plus the Waker queue internally, have the future do `try_borrow(_mut)` to attempt to "lock", and get basically all the performance you'd want.

Comment (scottbot95) [reply screeps-async/src/sync/mutex.rs:31]: Hmm that's a really good point actually
